modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/logical/LogicalOrToUnionRule.java

```diff
@@ -102,21 +96,100 @@ private LogicalOrToUnionRule(Config config) {
         if (operands == null)
             return;
 
+        if (!idxCollationCheck(call, operands))
+            return;
+
         RelNode input = getInput(call);
 
         RelNode rel0 = createUnionAll(cluster, input, operands.get(0), operands.get(1));
+        RelNode rel1 = createUnionAll(cluster, input, operands.get(1), operands.get(0));
 
-        call.transformTo(rel0, ImmutableMap.of(
-            createUnionAll(cluster, input, operands.get(1), operands.get(0)), rel0
-        ));
+        call.transformTo(rel0, ImmutableMap.of(rel1, rel0));
     }
 
     /** */
     protected abstract RexNode getCondition(RelOptRuleCall call);
 
+    /**
+     * Returns common required columns from scan.
+     */
+    protected ImmutableBitSet getRequiredColumns(RelOptRuleCall call, int fldCount) {
+        final IgniteLogicalTableScan scan = call.rel(0);
+
+        ImmutableBitSet.Builder builder = ImmutableBitSet.builder();
+
+        Mappings.TargetMapping mapping = Commons.inverseMapping(scan.requiredColumns(), fldCount);
+
+        new RexShuttle() {
+            @Override public RexNode visitLocalRef(RexLocalRef inputRef) {
+                builder.set(mapping.getSourceOpt(inputRef.getIndex()));
+                return inputRef;
+            }
+        }.apply(scan.condition());
+
+        builder.addAll(scan.requiredColumns());
```

Comment:
It's not correct, condition is built on top of scan.requiredColumns, why do we need to add it again? Moreover, we don't need to compute some "common required columns", it gives us nothing. We should compute it for each operands separately.